README.md
# autolibheif
autolibheif is a Linux CLI utility for convenient encoding and decoding of the HEIF/HEIC file format.

Supported formats: **.jpg / .png / .HEIF / .HEIC**
## Installation
### Arch-based distributions
Manual
```
git clone https://aur.archlinux.org/autolibheif.git
cd autolibheif/
makepkg -si
```
Using AUR helpers
```
yay -S autolibheif
```
### Other distributions
**Dependencies:**
[libheif](https://github.com/strukturag/libheif),
[python-setuptools](https://github.com/pypa/setuptools)

```
# create virtual environment if needed
git clone https://github.com/allddd/autolibheif.git
cd autolibheif/
python setup.py install
```
## Usage
Launch the utility from a terminal by typing `autolibheif`

change_log.txt
[X.X.X] XX/XX/XX
[IMPROVEMENT] improved performance
[BUG FIX] exit option in main menu throws an error

[1.0.2] 25/12/22
[IMPROVEMENT] instructions for each step by typing "h"
[IMPROVEMENT] important error messages now wait for user confirmation
[IMPROVEMENT] minor visual changes in the menus

[1.0.1] 09/12/22
[BUG FIX] encoding/decoding fails for files with uppercase extension


src/const.py
#
num_prompt = '''Type a number and confirm
==> '''
def_prompt = '''Type a number and confirm 
(leave blank for default)
==> '''
dir_prompt = '''
Type in full path to the photos folder and confirm
(leave blank to return to the main menu)
==> '''

#
undef_err = 'Something went wrong'
num_err = 'Please enter a valid number'
dir_err = 'Specified directory does not exist'
empty_err = 'Specified directory is empty'
nofiles_err = 'There are no files to convert in the specified directory'

#
row = '''
'''
enter = 'Press Enter to continue'
dir_txt = 'Looking for files in'
succ_txt = 'Successfully converted'
qty_txt = 'Select the desired conversion quality'

#
main_menu = ('''
[1] Decode (convert HEIC/HEIF file)
[2] Encode (convert file to HEIC/HEIF)
[3] Help
[4] Quit
''')
qty_menu = ('''
[1] Maximum (default)
[2] Decent
[3] Medium
[4] Low
[5] Exit to main menu
''')
dec_menu = ('''
[1] Convert to .jpg
[2] Convert to .png
[3] Exit to main menu
''')
enc_menu = ('''
[1] Convert to .HEIC
[2] Convert to .HEIF
[3] Exit to main menu
''')
help_menu = '''
[1] Exit to main menu (default)
[2] Visit the Github page
'''

#
help_main = ('''
Autolibheif is a Linux CLI utility for convenient
encoding and decoding of the HEIF/HEIC file format

If you need help at any time, type "h" and confirm
to get instructions for each individual step

Want to report a problem or request a feature ?
Select option [2] and visit the Github page
''')
help_decenc = ('''
Files will be converted to the file type you select here
''')
help_qty = ('''
The converted file retains a specific percentage of the 
original's quality, depending on what you select here

Maximum = 100%
Decent = 85%
Medium = 50%
Low = 25%
''')
help_dir = ('''
All compatible files in the folder specified
here will be converted to the desired file type

Other files and originals will remain intact

Enter the folder path in this format:
/home/user/Pictures
''')

#
github = 'https://github.com/allddd/autolibheif'
head = '''
 ▄▀▄ █ █ ▀█▀ ▄▀▄ █   █ ██▄ █▄█ ██▀ █ █▀
 █▀█ ▀▄█  █  ▀▄▀ █▄▄ █ █▄█ █ █ █▄▄ █ █▀
                                 v1.0.2'''


src/autolibheif.py
import os
import sys
import time
import subprocess
import webbrowser
from .const import *


dec_ext = ('.heif', '.HEIF', '.heic', '.HEIC')
enc_ext = ('.jpg', '.JPG', '.png', '.PNG')


# Menus
def main():
    while True:

        clear()
        print(head)
        print(main_menu)

        usr_inp = input(num_prompt)

        if usr_inp == '1':
            decode(ext=decodemenu(), qty=decqty(), dirr=decdir())
            break

        elif usr_inp == '2':
            encode(ext=encodemenu(), qty=encqty(), dirr=encdir())
            break

        elif usr_inp == '3':
            helpmenu()
            break

        elif usr_inp == 'h':
            helpmenu()
            break

        elif usr_inp == '4':
            clear()
            sys.exit(0)

        else:
            print(num_err)
            time.sleep(1)
            continue


def decodemenu():
    while True:

        clear()
        print(head)
        print(dec_menu)

        usr_inp = input(num_prompt)

        if usr_inp == '1':
            ext = '.jpg'
            return ext

        elif usr_inp == '2':
            ext = '.png'
            return ext

        elif usr_inp == '3':
            main()
            break

        elif usr_inp == 'h':
            print(help_decenc)
            ent()
            continue

        else:
            print(num_err)
            time.sleep(1)
            continue


def encodemenu():
    while True:

        clear()
        print(head)
        print(enc_menu)

        usr_inp = input(num_prompt)

        if usr_inp == '1':
            ext = '.HEIC'
            return ext

        elif usr_inp == '2':
            ext = '.HEIF'
            return ext

        elif usr_inp == '3':
            main()
            break

        elif usr_inp == 'h':
            print(help_decenc)
            ent()
            continue

        else:
            print(num_err)
            time.sleep(1)
            continue


def helpmenu():
    while True:

        clear()
        print(head)
        print(help_main)
        print(help_menu)

        usr_inp = input(def_prompt)

        if (not usr_inp):
            main()
            break

        elif usr_inp == '1':
            main()
            break

        elif usr_inp == '2':
            webbrowser.open(github)
            main()
            break

        else:
            print(num_err)
            time.sleep(1)
            continue


# Quality select
def decqty():
    while True:

        clear()
        print(head)
        print(f'{row}{qty_txt}')
        print(qty_menu)

        usr_inp = input(def_prompt)

        if (not usr_inp):
            qty = '100'
            return qty

        elif usr_inp == '1':
            qty = '100'
            return qty

        elif usr_inp == '2':
            qty = '85'
            return qty

        elif usr_inp == '3':
            qty = '50'
            return qty

        elif usr_inp == '4':
            qty = '25'
            return qty

        elif usr_inp == '5':
            main()
            break

        elif usr_inp == 'h':
            print(help_qty)
            ent()
            continue

        else:
            print(num_err)
            time.sleep(1)
            continue


def encqty():
    while True:

        clear()
        print(head)
        print(f'{row}{qty_txt}')
        print(qty_menu)

        usr_inp = input(def_prompt)

        if (not usr_inp):
            qty = '-L'
            return qty

        elif usr_inp == '1':
            qty = '-L'
            return qty

        elif usr_inp == '2':
            qty = '85'
            return qty

        elif usr_inp == '3':
            qty = '50'
            return qty

        elif usr_inp == '4':
            qty = '25'
            return qty

        elif usr_inp == '5':
            main()
            break

        elif usr_inp == 'h':
            print(help_qty)
            ent()
            continue

        else:
            print(num_err)
            time.sleep(1)
            continue


# Directory select
def decdir():
    while True:

        clear()
        print(head)
        dirr = input(dir_prompt)

        if (not dirr):
            main()
            break

        elif dirr == 'h':
            print(help_dir)
            ent()
            continue

        elif os.path.isdir(dirr):
            print(f'{row}{dir_txt} {dirr}{row}')

            if len(os.listdir(dirr)) == 0:
                print(empty_err)
                ent()
                continue

            else:
                for file in os.listdir(dirr):
                    if file.endswith(dec_ext):
                        return dirr

                else:
                    print(nofiles_err)
                    ent()
                    continue

        else:
            print(f'{row}{dir_err}')
            ent()
            continue


def encdir():
    while True:

        clear()
        print(head)
        dirr = input(dir_prompt)

        if (not dirr):
            main()
            break

        elif dirr == 'h':
            print(help_dir)
            ent()
            continue

        elif os.path.isdir(dirr):
            print(f'{row}{dir_txt} {dirr}{row}')

            if len(os.listdir(dirr)) == 0:
                print(empty_err)
                ent()
                continue

            else:
                for file in os.listdir(dirr):
                    if file.endswith(enc_ext):
                        return dirr

                else:
                    print(nofiles_err)
                    ent()
                    continue

        else:
            print(f'{row}{dir_err}')
            ent()
            continue


# Converting
def decode(ext, qty, dirr):
    cwd = os.getcwd()
    os.chdir(dirr)
    fnum = len(os.listdir())

    for file in os.listdir():
        if file.endswith(dec_ext):
            name = file[:-5]
            print(f'Converting {file} to {name}{ext}')
            subprocess.run(['heif-convert', '-q', qty, file, name + ext],
                           stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)

    else:
        new_fnum = len(os.listdir())
        conv_num = new_fnum - fnum

        if conv_num > 0:
            print(f'{row}{succ_txt} {conv_num} files')
            os.chdir(cwd)
            ent()
            main()

        else:
            print(undef_err)
            ent()
            os.chdir(cwd)
            main()


def encode(ext, qty, dirr):
    cwd = os.getcwd()
    os.chdir(dirr)
    fnum = len(os.listdir())

    for file in os.listdir():
        if file.endswith(enc_ext):
            name = file[:-4]
            print(f'Converting {file} to {name}{ext}')
            subprocess.run(['heif-enc', '-q', qty, file, '-o', name + ext],
                           stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)

    else:
        new_fnum = len(os.listdir())
        conv_num = new_fnum - fnum

        if conv_num > 0:
            print(f'{row}{succ_txt} {conv_num} files')
            os.chdir(cwd)
            ent()
            main()

        else:
            print(undef_err)
            ent()
            os.chdir(cwd)
            main()


# Other functions
def clear():
    os.system('clear')


def ent():
    print(enter)
    os.system('read -s')
